Sources/DistributedSystem/ChannelCompression.swift

```diff
@@ -0,0 +1,202 @@
+        buffer.writeInteger(compression, as: UInt8.self)
+        context.writeAndFlush(NIOAny(buffer), promise: nil)
+        if compression != 0 {
+            _ = context.pipeline.addHandler(ChannelCompressionInboundHandler(loggerBox), position: .after(streamHandler))
+            _ = context.pipeline.addHandler(ChannelCompressionOutboundHandler(), position: .last)
+        }
+        context.fireChannelActive()
+        _ = context.pipeline.removeHandler(self)
+    }
+}
+
+fileprivate struct BufferManager {
+    var buffers = Array(repeating: UnsafeMutableRawBufferPointer(start: nil, count: 0), count: 2)
+    var nextIdx = 0
+
+    mutating func getNextBuffer(capacity: Int) -> UnsafeMutableRawBufferPointer {
+        let idx = nextIdx
+        nextIdx = (1 - nextIdx)
+        var buffer = buffers[idx]
+        if buffer.count < capacity {
+            buffer.deallocate()
+            buffer = UnsafeMutableRawBufferPointer.allocate(byteCount: capacity, alignment: 0)
+            buffers[idx] = buffer
+        }
+        return buffer
+    }
+
+    mutating func reset() {
+        for idx in 0..<buffers.count {
+            buffers[idx].deallocate()
+        }
+    }
+}
+
+final class ChannelCompressionInboundHandler: ChannelInboundHandler {
+    typealias InboundIn = ByteBuffer
+    typealias InboundOut = ByteBuffer
+
+    private let loggerBox: Box<Logger>
+    private var bufferManager = BufferManager()
+    private var lz4Stream = LZ4_streamDecode_t()
+
+    private var logger: Logger { loggerBox.value }
+
+    init(_ loggerBox: Box<Logger>) {
+        self.loggerBox = loggerBox
+        LZ4_setStreamDecode(&lz4Stream, nil, 0);
+    }
+
+    deinit {
+        bufferManager.reset()
+    }
+
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        var bufferIn = unwrapInboundIn(data)
+        _ = bufferIn.readInteger(as: UInt32.self) // skip message size
+        if let messageSize = bufferIn.readInteger(as: UInt32.self) {
+            let dstBuffer = bufferManager.getNextBuffer(capacity: Int(messageSize))
+            let decompressedBytes = bufferIn.withUnsafeReadableBytes { ptr in
+                LZ4_decompress_safe_continue(&lz4Stream,
+                                             ptr.baseAddress!.assumingMemoryBound(to: Int8.self),
+                                             dstBuffer.baseAddress!.assumingMemoryBound(to: Int8.self),
+                                             Int32(ptr.count),
+                                             Int32(dstBuffer.count))
+            }
+            if decompressedBytes == messageSize {
+                var bufferOut = ByteBufferAllocator().buffer(capacity: MemoryLayout<UInt32>.size + Int(messageSize))
+                bufferOut.writeInteger(messageSize)
+                _ = bufferOut.writeWithUnsafeMutableBytes(minimumWritableBytes: 0) { ptr in
+                    ptr.copyMemory(from: UnsafeRawBufferPointer(start: dstBuffer.baseAddress, count: Int(messageSize)))
```

Comment:
Why can't we decompress directly to `bufferOut`?

Reply:
It is a specific of the LZ4 library.
On the compressor side we should preserve the data we compressed at the same memory address until the next compression call. The same for the decompressor side, we should preserve uncompressed data the same way.

Reply:
Maybe encapsulate this knowledge in e.g. `LZ4Compressor` that holds LZ4 stream and 2 buffers and have `compress/decompress()`  members like:
```
struct LZ4Compressor: ~Copyable {
    // stream and buffers
    func compress(_ inputBuffer: ByteBuffer, closure: (UnsafeRawBufferPointer) -> Void) {}

    func decompress(_ inputBuffer: ByteBuffer, closure: (UnsafeRawBufferPointer) -> Void) {}

}
```
 [](edited-with-Axolo)